lambda/shared/src/main/scala/feral/lambda/events/S3BatchResult.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.events
+
+import io.circe.Encoder
+
+// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-lambda/trigger/s3-batch.d.ts
+
+final case class S3BatchResult(
+    invocationSchemaVersion: String,
+    treatMissingKeysAs: S3BatchResultCode,
+    invocationId: String,
+    results: List[S3BatchResultResult]
+)
+
+object S3BatchResult {
+  implicit val encoder: Encoder[S3BatchResult] =
+    Encoder.forProduct4(
+      "invocationSchemaVersion",
+      "treatMissingKeysAs",
+      "invocationId",
+      "results")(r =>
+      (r.invocationSchemaVersion, r.treatMissingKeysAs, r.invocationId, r.results))
+}
+
+sealed trait S3BatchResultCode
```

Comment:
Admittedly, it's kind of a crazy name 😂 

```suggestion
sealed abstract class S3BatchResultResultCode
```
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d18a30127146217038c61060b06e96713daa706c/types/aws-lambda/trigger/s3-batch.d.ts#L35

Reply:
No problem. Any reason why `abstract class` works better than `trait` here?

Reply:
Good question, in practice it will probably make no measurable difference. I prefer `abstract class` where possible because admittedly, superstition, 😂 I think it _might_ be more performant to do `isInstanceOf` checks and stuff.

More interestingly, the new `enum` keyword in Scala 3 would let us define it like this as:
```
enum S3BatchResultResultCode:
  case Succeeded, TemporaryFailure, PermanentFailure
```

And behind-the-scenes, that is generating code equivalent to using an `abstract class`. So this approach is binary compatible with that syntax sugar.

Reply:
Gotcha, makes sense.